lib/std/crypto/salsa20.zig

```diff
@@ -0,0 +1,430 @@
+        Salsa20.xor(out, in, counter, extended.key, extended.nonce);
+    }
+};
+
+/// The XSalsa20 stream cipher, combined with the Poly1305 MAC
+pub const XSalsa20Poly1305 = struct {
+    /// Authentication tag length in bytes.
+    pub const tag_length = Poly1305.mac_length;
+    /// Nonce length in bytes.
+    pub const nonce_length = XSalsa20.nonce_length;
+    /// Key length in bytes.
+    pub const key_length = XSalsa20.key_length;
+
+    /// c: ciphertext: output buffer should be of size m.len
+    /// tag: authentication tag: output MAC
+    /// m: message
+    /// ad: Associated Data
+    /// npub: public nonce
+    /// k: private key
+    pub fn encrypt(c: []u8, tag: *[tag_length]u8, m: []const u8, ad: []const u8, npub: [nonce_length]u8, k: [key_length]u8) void {
+        debug.assert(c.len == m.len);
+        const extended = extend(k, npub);
+        var block0 = [_]u8{0} ** 64;
+        const mlen0 = math.min(32, m.len);
+        mem.copy(u8, block0[32..][0..mlen0], m[0..mlen0]);
+        Salsa20.xor(block0[0..], block0[0..], 0, extended.key, extended.nonce);
+        mem.copy(u8, c[0..mlen0], block0[32..][0..mlen0]);
+        Salsa20.xor(c[mlen0..], m[mlen0..], 1, extended.key, extended.nonce);
+        var mac = Poly1305.init(block0[0..32]);
+        mac.update(ad);
+        mac.update(c);
+        mac.final(tag);
+    }
+
+    /// m: message: output buffer should be of size c.len
+    /// c: ciphertext
+    /// tag: authentication tag
+    /// ad: Associated Data
+    /// npub: public nonce
+    /// k: private key
+    pub fn decrypt(m: []u8, c: []const u8, tag: [tag_length]u8, ad: []const u8, npub: [nonce_length]u8, k: [key_length]u8) !void {
+        debug.assert(c.len == m.len);
+        const extended = extend(k, npub);
+        var block0 = [_]u8{0} ** 64;
+        const mlen0 = math.min(32, c.len);
+        mem.copy(u8, block0[32..][0..mlen0], c[0..mlen0]);
+        Salsa20.xor(block0[0..], block0[0..], 0, extended.key, extended.nonce);
+        var mac = Poly1305.init(block0[0..32]);
+        mac.update(ad);
+        mac.update(c);
+        var computedTag: [tag_length]u8 = undefined;
+        mac.final(&computedTag);
+        var acc: u8 = 0;
+        for (computedTag) |_, i| {
+            acc |= (computedTag[i] ^ tag[i]);
+        }
+        if (acc != 0) {
+            mem.secureZero(u8, &computedTag);
+            return error.AuthenticationFailed;
+        }
+        mem.copy(u8, m[0..mlen0], block0[32..][0..mlen0]);
+        Salsa20.xor(m[mlen0..], c[mlen0..], 1, extended.key, extended.nonce);
+    }
+};
+
+/// NaCl-compatible secretbox API.
+///
+/// A secretbox contains both an encrypted message and an authentication tag to verify that it hasn't been tampered with.
+/// A secret key shared by all the recipients must be already known in order to use this API.
+///
+/// Nonces are 192-bit large and can safely be chosen with a random number generator.
+pub const secretBox = struct {
```

Comment:
This and others should be Title Case

Reply:
Right! I'll send a following PR to change this, thanks!

Reply:
Thanks both of you :)